pkg/registration/spoke/registration/hub_timeout_controller.go

```diff
@@ -0,0 +1,70 @@
+package registration
+
+import (
+	"context"
+	"time"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	coordlisters "k8s.io/client-go/listers/coordination/v1"
+	"k8s.io/klog/v2"
+)
+
+const leaseName = "managed-cluster-lease"
+
+type hubTimeoutController struct {
+	clusterName        string
+	leaseLister        coordlisters.LeaseLister
+	timeoutSeconds     int32
+	lastLeaseRenewTime time.Time
+	handleTimeout      func(ctx context.Context) error
+}
+
+func NewHubTimeoutController(
+	clusterName string,
+	leaseLister coordlisters.LeaseLister,
+	timeoutSeconds int32,
+	handleTimeout func(ctx context.Context) error,
+	recorder events.Recorder,
+) factory.Controller {
+	c := &hubTimeoutController{
+		clusterName:    clusterName,
+		timeoutSeconds: timeoutSeconds,
+		handleTimeout:  handleTimeout,
+		leaseLister:    leaseLister,
+	}
+	return factory.New().WithSync(c.sync).ResyncEvery(time.Minute).
+		ToController("HubTimeoutController", recorder)
+}
+
+func (c *hubTimeoutController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	logger := klog.FromContext(ctx)
+	if c.handleTimeout == nil {
+		return nil
+	}
+
+	lease, err := c.leaseLister.Leases(c.clusterName).Get(leaseName)
```

Comment:
Can we just get lease using client? user lister does not make sense here, it works only when you watch instead of polling.